core/src/main/java/com/linecorp/armeria/client/ClientConnectionTimings.java

```diff
@@ -0,0 +1,248 @@
+     * Returns the duration which was taken to resolve a domain name, in nanoseconds.
+     *
+     * @return the duration, or {@code -1} if there was no action to resolve a domain name.
+     */
+    public long dnsResolutionDurationNanos() {
+        return dnsResolutionDurationNanos;
+    }
+
+    /**
+     * Returns the time when connecting to a remote peer started, in microseconds since the epoch.
+     *
+     * @return the duration, or {@code -1} if there was no action to connect to a remote peer.
+     */
+    public long socketConnectStartTimeMicros() {
+        return socketConnectStartTimeMicros;
+    }
+
+    /**
+     * Returns the time when connecting to a remote peer started, in milliseconds since the epoch.
+     *
+     * @return the duration, or {@code -1} if there was no action to connect to a remote peer.
+     */
+    public long socketConnectStartTimeMillis() {
+        if (socketConnectStartTimeMicros >= 0) {
+            return TimeUnit.MICROSECONDS.toMillis(socketConnectStartTimeMicros);
+        }
+        return -1;
+    }
+
+    /**
+     * Returns the duration which was taken to connect to a remote peer, in nanoseconds.
+     *
+     * @return the duration, or {@code -1} if there was no action to connect to a remote peer.
+     */
+    public long socketConnectDurationNanos() {
+        return socketConnectDurationNanos;
+    }
+
+    /**
+     * Returns the time when waiting the completion of an existing connection attempt started,
+     * in microseconds since the epoch.
+     *
+     * @return the duration, or {@code -1} if there was no action to get a pending connection.
+     */
+    public long pendingAcquisitionStartTimeMicros() {
+        return pendingAcquisitionStartTimeMicros;
+    }
+
+    /**
+     * Returns the time when waiting the completion of an existing connection attempt started,
+     * in milliseconds since the epoch.
+     *
+     * @return the duration, or {@code -1} if there was no action to get a pending connection.
+     */
+    public long pendingAcquisitionStartTimeMillis() {
+        if (pendingAcquisitionStartTimeMicros >= 0) {
+            return TimeUnit.MICROSECONDS.toMillis(pendingAcquisitionStartTimeMicros);
+        }
+        return -1;
+    }
+
+    /**
+     * Returns the duration which was taken to wait the completion of an existing connection attempt in order to
+     * use one connection for HTTP/2.
+     *
+     * @return the duration, or {@code -1} if there was no action to get a pending connection.
+     */
+    public long pendingAcquisitionDurationNanos() {
+        return pendingAcquisitionDurationNanos;
+    }
+
+    @Override
+    public String toString() {
+        final ToStringHelper toStringHelper =
+                MoreObjects.toStringHelper(this)
+                           .add("connectionAcquisitionStartTimeMicros", connectionAcquisitionStartTimeMicros)
+                           .add("connectionAcquisitionDurationNanos", connectionAcquisitionDurationNanos);
```

Comment:
How about making use of `c.l.a.common.util.TextFormatter`? You'll have to add two new methods though:

- `epochMicros(long timeMicros)`
- `appendEpochMicros(StringBuilder buf, long timeMicros)`

Durations could be formatted with `elapsed(long)` easily.

Reply:
That's a good suggestion. Added it.